crates/compiler/builtins/roc/Json.roc

```diff
@@ -522,123 +486,752 @@ tryDecode = \{ result, rest }, mapper ->
+
+                        # Build final record from decoded fields and values
+                        when finalizer updatedRecord is 
+                            Ok val -> { result: Ok val, rest }
+                            Err e -> { result: Err e, rest }
+
+                    _ ->
+                        # Invalid object 
+                        {result : Err TooShort, rest: bytesAfterValue}
+            
+         
+    countBytesBeforeFirstField = 
+        when List.walkUntil bytes (BeforeOpeningBrace 0) objectHelp is 
+            ObjectFieldNameStart n -> n
+            _ -> 0
+
+    if countBytesBeforeFirstField == 0 then 
+        # Invalid object, expected opening brace '{' followed by a field
+        {result: Err TooShort, rest: bytes}
+    else
+        bytesBeforeFirstField = (List.drop bytes countBytesBeforeFirstField)
+
+        # Begin decoding field:value pairs
+        decodeFields initialState bytesBeforeFirstField
+
+objectHelp : ObjectState, U8 -> [Break ObjectState, Continue ObjectState]
+objectHelp = \state, byte ->
+    when (state, byte) is 
+        (BeforeOpeningBrace n, b) if isWhitespace b -> Continue (BeforeOpeningBrace (n+1))
+        (BeforeOpeningBrace n, b) if b == '{' -> Continue (AfterOpeningBrace (n+1))
+        (AfterOpeningBrace n, b) if isWhitespace b -> Continue (AfterOpeningBrace (n+1))
+        (AfterOpeningBrace n, b) if b == '"' -> Break (ObjectFieldNameStart n) # field names must be a json string
+        (BeforeColon n, b) if isWhitespace b -> Continue (BeforeColon (n+1))
+        (BeforeColon n, b) if b == ':' -> Continue (AfterColon (n+1))
+        (AfterColon n, b) if isWhitespace b -> Continue (AfterColon (n+1))
+        (AfterColon n, _) -> Break (ObjectValueStart n) # object value could start with **almost** anything
+        (AfterObjectValue n, b) if isWhitespace b -> Continue (AfterObjectValue (n+1))
+        (AfterObjectValue n, b) if b == ',' -> Continue (AfterComma (n+1))
+        (AfterObjectValue n, b) if b == '}' -> Continue (AfterClosingBrace (n+1))
+        (AfterComma n, b) if isWhitespace b -> Continue (AfterComma (n+1))
+        (AfterComma n, b) if b == '"' -> Break (ObjectFieldNameStart n)
+        (AfterClosingBrace n, b) if isWhitespace b -> Continue (AfterClosingBrace (n+1))
+        _ -> Break InvalidObject
+
+ObjectState : [
+    BeforeOpeningBrace Nat,
+    AfterOpeningBrace Nat,
+    ObjectFieldNameStart Nat,
+    BeforeColon Nat, 
+    AfterColon Nat, 
+    ObjectValueStart Nat, 
+    AfterObjectValue Nat,
+    AfterComma Nat,
+    AfterClosingBrace Nat, 
+    InvalidObject,
+]
+
+# Test decode of simple Json Object into Roc Record
+expect
+    input = Str.toUtf8 "{\"fruit\":2\n} "
+
+    actual = Decode.fromBytesPartial input fromUtf8
+    expected = Ok { fruit: 2 }
+
+    actual.result == expected
+
+# Test decode of simple Json Object into Roc Record
+# TODO assertion failed: value == 0 || value == 1
```

Comment:
@ayazhafiz @folkertdev I'm also getting that assert fail here. I'm reasonably sure I should be able to track it down. But thought you may be interested.

Reply:
I found some more interesting information. I summarised in this [gist](https://gist.github.com/lukewilliamboswell/5f2cf3f472324fa6e73b75d01bb8f238). Essentially, I am convinced that the decoders are interfering which each other somehow and causing strange behaviour.

Reply:
I merged in main and have done some further testing. I can confirm that #5210 fixed the assert bug. However the runtime bug remains. Running the below tests for `decodeRecord` somehow interferes with the other decoder tests causing them to fail, these tests will pass but those other tests fail (which were previously passing).

```elixir
# Test decode of simple Json Object into Roc Record ignoring whitespace
# TODO this test makes `decodeI64` test fail
expect
    input = Str.toUtf8 " {\n\"fruit\"\t:2\n } "

    actual = Decode.fromBytesPartial input fromUtf8
    expected = Ok { fruit: 2 }

    actual.result == expected

# Test decode of simple Json Object
# TODO this test makes `decodeString` tests fail
expect
    input = Str.toUtf8 "{\"fruit\": \"apple\" }"

    actual = Decode.fromBytesPartial input fromUtf8
    expected = Ok { fruit: "apple" }

    actual.result == expected

# Test decode of simple Json Object
# TODO this test makes `decodeBool` tests fail
expect
    input = Str.toUtf8 "{\"fruit\": true }"

    actual : DecodeResult { fruit : Bool }
    actual = Decode.fromBytesPartial input fromUtf8
    expected = Ok { fruit: Bool.true }

    actual.result == expected
```

Reply:
Workaround added by including another element in record. Thank you @ayazhafiz